core/src/test/java/com/graphhopper/routing/weighting/custom/SpeedCalculatorTest.java

```diff
@@ -0,0 +1,180 @@
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.graphhopper.routing.weighting.custom;
+
+import com.graphhopper.routing.profiles.EnumEncodedValue;
+import com.graphhopper.routing.profiles.RoadClass;
+import com.graphhopper.routing.profiles.RoadEnvironment;
+import com.graphhopper.routing.util.CustomModel;
+import com.graphhopper.routing.util.EncodingManager;
+import com.graphhopper.routing.util.FlagEncoder;
+import com.graphhopper.storage.GraphBuilder;
+import com.graphhopper.util.EdgeIteratorState;
+import org.junit.jupiter.api.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class SpeedCalculatorTest {
+
+    private final EncodingManager em;
+    private final EdgeIteratorState edge;
+
+    SpeedCalculatorTest() {
+        // in this test we use the same edge for every test and check the calculated speed for different custom models
+        // and depending on additional properties of the edge we set in the tests
+        em = EncodingManager.create("car");
+        edge = new GraphBuilder(em).create().edge(0, 1, 1000, true);
+    }
+
+    @Test
+    public void maxSpeedFallback() {
+        // default speed for car is 60km/h
+        assertEquals(60, calcSpeed(edge, new CustomModel()));
+        // use global fallback to limit speed
+        assertEquals(30, calcSpeed(edge, new CustomModel().setMaxSpeedFallback(30.0)));
+    }
+
+    @Test
+    public void maxSpeed() {
+        // here we use max_speed to limit speed
+        CustomModel model = new CustomModel();
+        Map<String, Object> map = new HashMap<>();
+        map.put("*", 25);
+        model.getMaxSpeed().put(RoadClass.KEY, map);
+        assertEquals(25, calcSpeed(edge, model));
+        // ... but setting it to a higher value than the default has no effect
+        map.put("*", 70);
+        assertEquals(60, calcSpeed(edge, model));
+        // also our edge might or might not be affected
+        // todonow: why has encoding manager a road class encoded value by default?
+        EnumEncodedValue<RoadClass> roadClass = em.getEnumEncodedValue(RoadClass.KEY, RoadClass.class);
+        edge.set(roadClass, RoadClass.PRIMARY);
+        // todonow: why do we not allow an empty map as value?
```

Comment:
An empty map would be the same as not specifying any rules for this key, is there a problem with this?

Reply:
An empty map means we set something and then left it empty which indicates some mistake. But yes, if you do not like it we can allow it.

Reply:
Why is this a mistake? I thought maybe I do `speed_factor: road_class: {motorway: 0.3}` and then I want to remove it and just change it to `speed_factor: road_class: {}`.. or could this happen another way?

Reply:
What exactly is the use case here? With "removing" do you mean "setting motorway back to 1" of a server-side custom model with a query? (this won't be allowed for LM :) )

Reply:
No, I was just thinking about editing the file. If I have `road_class: {motorway: 0.3}` and the next day I do not like it anymore I have to remove the `road_class` key and cannot simply put `road_class: {}`. Nothing major really we can also leave it as it is.

Reply:
Ah, ok. We can leave it as it is or allow it (have no strong opinion)

Reply:
Will leave as is